### PR TITLE
vSphere provider for Skypilot

### PR DESCRIPTION
**Pull request by Decikingship**
vSphere provider for Skypilot

Co-authored-by: Yixing Jia <yixingjia@gmail.com>
Co-authored-by: YongChun Chang <changsang91@163.com>
Co-authored-by: Shiqi He <johnhsq@126.com>

Sign-off-by: Shiqi He <johnhsq@126.com>

### CODE REVIEW

File: sky/clouds/service_catalog/common.py [hunk 1]
```diff
@@ -487,6 +487,10 @@ def list_accelerators_impl(
         'SpotPrice',
         'Region',
     ]].dropna(subset=['AcceleratorName']).drop_duplicates()
+    # In case of the df is empty, it should return an empty array
```
Comment (yixingjia) [inline sky/clouds/service_catalog/common.py:490]: Please revert this change.

File: sky/setup_files/MANIFEST.in [hunk 2]
```diff
@@ -11,6 +11,7 @@ include sky/skylet/providers/ibm/*
 include sky/skylet/providers/kubernetes/*
 include sky/skylet/providers/lambda_cloud/*
 include sky/skylet/providers/oci/*
+include sky/skylet/providers/vsphere/*
```
[the remark this answers is not in the document]
Comment (Decikingship) [reply sky/setup_files/MANIFEST.in]: removed

File: sky/clouds/service_catalog/common.py (hunk 1 above)
Comment (Decikingship) [reply sky/clouds/service_catalog/common.py]: removed

File: sky/clouds/vsphere.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+        region: Optional[str],
+        zone: Optional[str],
+    ) -> List[clouds.Region]:
+        del accelerators, zone  # unused
+        if use_spot:
+            return []
+        if instance_type is None:
+            # Fall back to default regions
+            regions = cls.regions()
+        else:
```
Comment (Michaelvll) [inline sky/clouds/vsphere.py:92]: The `instance_type` should be guaranteed to be specified by the optimizer, we don't have to do the if condition here.

File: sky/provision/vsphere/common/cls_api_client.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+from com.vmware.content.library_client import Subscriptions
+from com.vmware.content_client import Library
+from com.vmware.content_client import LocalLibrary
+from com.vmware.content_client import SubscribedLibrary
+from com.vmware.vcenter.iso_client import Image
+from com.vmware.vcenter.ovf_client import LibraryItem
+from com.vmware.vcenter.vm_template.library_items_client import CheckOuts
+from com.vmware.vcenter.vm_template.library_items_client import Versions
+from com.vmware.vcenter.vm_template_client import LibraryItems
+from com.vmware.vcenter_client import VM
```
Comment (Michaelvll) [inline sky/provision/vsphere/common/cls_api_client.py:22]: For these imports, we should move them to the `sky/adaptors`, i.e., it should not be directly imported, but only be imported when being used. Reason: for a user without `vmware` dependencies installed, they will experience an `ImportError` as python will import these modules when `__init__` exists.

File: sky/provision/vsphere/instance.py [hunk 5]
```diff
@@ -0,0 +1,604 @@
...
+from pyVmomi import vim  # type: ignore
+
+from sky import exceptions
+from sky import sky_logging
+from sky import status_lib
+from sky.clouds.service_catalog.common import get_catalog_path
+from sky.provision import common
+from sky.provision.vsphere import vsphere_utils
+from sky.provision.vsphere.common.custom_script import CUSTOMIZED_SCRIPT
+from sky.provision.vsphere.common.metadata_utils import Metadata
```
Comment (Michaelvll) [inline sky/provision/vsphere/instance.py:17]: style nit: we normally import modules only, not the classes directly. Reference: https://google.github.io/styleguide/pyguide.html#22-imports

File: sky/setup_files/setup.py [hunk 6]
```diff
@@ -233,6 +234,10 @@ def parse_readme(readme: str) -> str:
     'oci': ['oci'] + local_ray,
     'kubernetes': ['kubernetes>=20.0.0'] + local_ray,
     'remote': remote,
+    'vsphere': [
+        'pyvmomi==8.0.1.0.2',
+        'vsphere-automation-sdk @ git+https://github.com/vmware/vsphere-automation-sdk-python.git@v8.0.1.0'
+    ] + local_ray,
```
Comment (Michaelvll) [inline sky/setup_files/setup.py:240]: `local_ray` should not be required as we are not using ray autoscaler locally.
```suggestion
    ],
```

File: sky/templates/vsphere-ray.yml.j2 [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+idle_timeout_minutes: 60
+
+provider:
+  type: external
+  module: sky.provision.vsphere
+  region: {{region}}
+  availability_zone: {{zones}}
+
+auth:
+  ssh_user: ecl
```
Comment (Michaelvll) [inline sky/templates/vsphere-ray.yml.j2:15]: nit: Do we want to make it a more general name, such as `ubuntu` or `vsphereuser`?

File: sky/templates/vsphere-ray.yml.j2 [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    node_config:
+      InstanceType: {{instance_type}}
+{% if num_nodes > 1 %}
+  ray_worker_default:
+    min_workers: {{num_nodes - 1}}
+    max_workers: {{num_nodes - 1}}
+    resources: {}
+    node_config:
+      InstanceType: {{instance_type}}
+{%- endif %}
```
Comment (Michaelvll) [inline sky/templates/vsphere-ray.yml.j2:30]: These part is no longer needed, as our new provisioner will not use the worker config.

File: sky/templates/vsphere-ray.yml.j2 [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    sudo kill -9 `sudo lsof /var/lib/dpkg/lock-frontend | awk '{print $2}' | tail -n 1` || true;
+    sudo pkill -9 apt-get;
+    sudo pkill -9 dpkg;
+    sudo dpkg --configure -a;
+    mkdir -p ~/.ssh; touch ~/.ssh/config;
+    pip3 --version > /dev/null 2>&1 || (curl -sSL https://bootstrap.pypa.io/get-pip.py -o get-pip.py && python3 get-pip.py && echo "PATH=$HOME/.local/bin:$PATH" >> ~/.bashrc);
+    (type -a python | grep -q python3) || echo 'alias python=python3' >> ~/.bashrc;
+    (type -a pip | grep -q pip3) || echo 'alias pip=pip3' >> ~/.bashrc;
+    which conda > /dev/null 2>&1 || (wget -nc https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh && bash Miniconda3-latest-Linux-x86_64.sh -b && eval "$(~/miniconda3/bin/conda shell.bash hook)" && conda init && conda config --set auto_activate_base true);
+    source ~/.bashrc;
```
Comment (Michaelvll) [inline sky/templates/vsphere-ray.yml.j2:74]: For conda installation, please check the conda installation commands in other clouds: https://github.com/skypilot-org/skypilot/blob/9743aa0c7f86a352715b13dd18ee780239f190f6/sky/templates/aws-ray.yml.j2#L158

We don't install the latest conda version, as Python3.11 is not officially supported at the moment.

File: sky/templates/vsphere-ray.yml.j2 [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    which prlimit && for id in $(pgrep -f raylet/raylet); do sudo prlimit --nofile=1048576:1048576 --pid=$id || true; done;
+{%- else %}
+worker_start_ray_commands: []
+{%- endif %}
+
+# These fields are required for external cloud providers.
+head_setup_commands: []
+worker_setup_commands: []
+cluster_synced_files: []
+file_mounts_sync_continuously: False
```
Comment (Michaelvll) [inline sky/templates/vsphere-ray.yml.j2:108]: These lines are no longer needed considering that we are using the new provisioner to setup the runtime for the instances.

https://github.com/skypilot-org/skypilot/blob/9743aa0c7f86a352715b13dd18ee780239f190f6/sky/templates/aws-ray.yml.j2#L172-L174

**Review by Michaelvll — COMMENTED** [review]
Thank you for submitting the PR for support Vsphere @Decikingship! We are very excited to see the support. Left several comments. Most of the code looks good to me.
One thing that we need to be careful is that a user without vsphere credentials and [vsphere] dependencies installed should be able to use the other clouds normally. Could you help double check that?

File: sky/clouds/vsphere.py [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+            clouds=_CLOUD_VSPHERE,
+        )
+        if instance_list is None:
+            return ([], fuzzy_candidate_list)
+        return (_make(instance_list), fuzzy_candidate_list)
+
+    @classmethod
+    def check_credentials(cls) -> Tuple[bool, Optional[str]]:
+        """Checks if the user has access credentials to this cloud."""
+
```
Comment (Michaelvll) [inline sky/clouds/vsphere.py:265]: It would be also good to check if the dependencies are correctly installed in this function, i.e. we should print out some hints if the dependencies are not correctly installed during `sky check`, e.g., https://github.com/skypilot-org/skypilot/blob/9743aa0c7f86a352715b13dd18ee780239f190f6/sky/clouds/aws.py#L464-L486

File: sky/provision/vsphere/common/cls_api_client.py (hunk 4 above)
Comment (Decikingship) [reply sky/provision/vsphere/common/cls_api_client.py:22]: Should we convert all files in sky/provision/vsphere/common into methods inside adaptors/vsphere.py in this case? Or simply move imports inside the functions?
I need clarification because files in sky/provision/vsphere/common were originally located in sky/skylet/providers/vsphere. We moved them because we were notified that sky/skylet/providers folder would be deprecated soon. Under sky/skylet/providers, imports like boto3 and googleapiclient were directly imported.

File: sky/clouds/service_catalog/vsphere_catalog.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+                '\'MemoryInfo\': {\'SizeInMiB\': 81559}}], '
+                '\'TotalGpuMemoryInMiB\': 81559}\",vmvare,0.01,\nvmvare-2,'
+                'H100-2-80G,48,480,H100,2,\"{\'Gpus\': [{\'Name\': \'NVIDIA '
+                'H100 PCIe\', \'Manufacturer\': \'NVIDIA\', \'Count\': 2, '
+                '\'MemoryInfo\': {\'SizeInMiB\': 81559}}], '
+                '\'TotalGpuMemoryInMiB\': 163118}\",vmvare,0.01,\nvmvare-2,'
+                'vsphere-S,10,42,P100,1,\"{\'Gpus\': [{\'Name\': \'Tesla '
+                'P100-PCIE-16GB\', \'Manufacturer\': \'NVIDIA\', \'Count\': '
+                '1, \'MemoryInfo\': {\'SizeInMiB\': 16384}}], '
+                '\'TotalGpuMemoryInMiB\': 16384}\",vmvare,0.01')
```
[the remark this answers is not in the document]
Comment (Decikingship) [reply sky/clouds/service_catalog/vsphere_catalog.py]: Have replaced this with monkeypatch in common.py and tests/default_vsphere/vms.csv

File: sky/clouds/vsphere.py [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+                f'not supported on {cls._REPR}.')
+        return features
+
+    @classmethod
+    def max_cluster_name_length(cls) -> Optional[int]:
+        return cls._MAX_CLUSTER_NAME_LEN_LIMIT
+
+    @classmethod
+    def regions(cls) -> List[clouds.Region]:
+        return service_catalog.regions(clouds=_CLOUD_VSPHERE)
```
[the remark this answers is not in the document]
Comment (Decikingship) [reply sky/clouds/vsphere.py]: Removed

File: sky/clouds/vsphere.py (hunk 3 above)
Comment (Decikingship) [reply sky/clouds/vsphere.py]: Removed

File: sky/setup_files/setup.py (hunk 6 above)
Comment (Decikingship) [reply sky/setup_files/setup.py]: Removed

File: sky/templates/vsphere-ray.yml.j2 (hunk 7 above)
Comment (Decikingship) [reply sky/templates/vsphere-ray.yml.j2]: Changed to `ubuntu`

File: sky/templates/vsphere-ray.yml.j2 (hunk 8 above)
Comment (Decikingship) [reply sky/templates/vsphere-ray.yml.j2]: Removed

File: sky/templates/vsphere-ray.yml.j2 (hunk 9 above)
Comment (Decikingship) [reply sky/templates/vsphere-ray.yml.j2]: Updated with ` {{ conda_installation_commands }} `

File: sky/templates/vsphere-ray.yml.j2 (hunk 10 above)
Comment (Decikingship) [reply sky/templates/vsphere-ray.yml.j2]: Removed

File: sky/provision/vsphere/common/cls_api_client.py (hunk 4 above)
Comment (Michaelvll) [reply sky/provision/vsphere/common/cls_api_client.py:22]: Moving `sky/provision/vsphere/common` to `adaptors/vsphere.py` will not directly work as we still have to handle the imports with the functions. I am fine with importing the modules inside the functions, or we can create a adaptor function in `adaptors/vsphere.py` that returns the `com.vmware`.

```python
vmware = None

def import_package(func):

    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        global vmware
        if vmware is None:
            try:
                import com.vmware as _vmware
            except ImportError:
                raise ImportError('Fail to import dependencies for Vsphere. '
                                  'Try pip install "skypilot[vsphere]"') from None
        return func(*args, **kwargs)

    return wrapper

@import_pacakge
def vmware():
    return _vmware
```

File: sky/clouds/vsphere.py (hunk 11 above)
Comment (Decikingship) [reply sky/clouds/vsphere.py:254]: Hints added

File: sky/provision/vsphere/common/cls_api_client.py (hunk 4 above)
Comment (Decikingship) [reply sky/provision/vsphere/common/cls_api_client.py]: Adaptors added

**Review by Michaelvll — APPROVED** [review]
Thanks for the great effort for supporting vsphere in SkyPilot @Decikingship! I just tested this PR without vsphere enabled, and it seems the new PR does not affect the other clouds.

Considering that we currently don't have the access to an available Vsphere cluster and the urgent timeline, it should be fine to merge this PR to the master first and improve in the future.

Once the style issues are fixed and the tests are passed internally. It should be good to go in.
Could you leave a comment for the tests you have done internally for vsphere, so we can better keep track of the current status of the PR? @Decikingship

File: sky/provision/vsphere/instance.py (hunk 5 above)
Comment (Decikingship) [reply sky/provision/vsphere/instance.py]: fixed

**Comment by Decikingship** [comment]
> Thanks for the great effort for supporting vsphere in SkyPilot @Decikingship! I just tested this PR without vsphere enabled, and it seems the new PR does not affect the other clouds.
> 
> Considering that we currently don't have the access to an available Vsphere cluster and the urgent timeline, it should be fine to merge this PR to the master first and improve in the future.
> 
> Once the style issues are fixed and the tests are passed internally. It should be good to go in. Could you leave a comment for the tests you have done internally for vsphere, so we can better keep track of the current status of the PR? @Decikingship

Hi @Michaelvll The tests we have done are:
sky check - passed
sky launch (CPU case) - passed
sky launch (GPU case) - passed
sky cancel - passed
sky stop - passed
sky down - passed
sky exec - passed
sky status - passed
sky queue - passed
sky logs - passed
sky start - passed
sky cpunode - passed
sky gpunode - passed
sky show-gpus - passed
sky storage - passed
sky admin - passed

sky spot - not supported
sky autostop - not supported
sky bench - not supported
sky tpunode - not supported
sky cost-report - not supported

**Comment by Michaelvll** [comment]
Thanks for listing out the tests we have done @Decikingship! I just run the following test without the vsphere enabled:
- [x] `pytest tests/test_smoke.py`
